### ISSUE TITLE
No LICENSE?

### ISSUE THREAD

**Issue by teeler** [blevesearch/snowballstem#2]
There's no LICENSE in this repo, would you please add one?

**Comment by pombredanne** [comment]
@teeler FWIW, based on what I can see these are all generated by the Snowball compiler at  https://snowballstem.org/

> The Snowball compiler translates a Snowball script into another language - currently ISO C, C#, Go, Java, Javascript, Object Pascal, Python and Rust are supported. 

The license is at https://snowballstem.org/license.html or https://github.com/snowballstem/snowball/blob/master/COPYING and is a BSD-3-Clause.
The Go backend was contributed by @mschoch https://github.com/snowballstem/snowball/commits?author=mschoch who as it happens is also the key and original author of bleve https://github.com/blevesearch/bleve

@mschoch do you need help to add this license?

There are a few discussions on the license of Snowball in general there https://www.google.com/search?q=site%3Ahttps%3A%2F%2Flists.tartarus.org%2Fpipermail%2Fsnowball-discuss%2F%20license&pws=0&gl=us&gws_rd=cr

See in particular this about clarifying the BSD-3-Clause vs. BSD-2-Clause https://lists.tartarus.org/pipermail/snowball-discuss/2015-September/001590.html

**Comment by teeler** [comment]
@pombredanne Thanks - I saw those licenses. I need an explicit license with this code, though, otherwise my compliance folks won't let me use it.

**Comment by mschoch** [comment]
I do not need help, I would need legal advice.  I will probably have to remove this code instead.

**Comment by willnorris** [comment]
@mschoch: wouldn't this simply inherit the BSD-3-Clause license from Snowball?  Or at least, is there any harm, or would you object, to putting your generator code here under the BSD-3-Clause (even though that differs from the Apache 2.0 used in Bleve itself)?

**Comment by mschoch** [comment]
No, that is not my understanding.  Typically the output of a program does not inherit the license of the program that generated it (with the obvious exception if the output includes substantial portions of the original).  See this thread, about GPL, where the implications are perhaps more important, but relevant here nonetheless.  https://opensource.stackexchange.com/questions/2501/does-using-a-gplv2-code-generator-force-my-project-to-be-similarly-licensed

I would agree that this is probably the case for the runtime support portion of the code, which was part of my contribution to the snowball project.  But, the part I'm more worried about is the actual language specific generated code.  To me, the snowball rule/data files themselves should have started with a clear license, and that more plausibly applies to the generated code (though even that can be debated).  But that's the real complication, my understand from researching this in the past is that many of those language files were contributed by external people and no clear license was established or enforced.

Maybe it would be possible to at least investigate this on per-language basis, as I suspect some of the do have clear history and license.

As for your other question, I don't think there is any "everything has to be Apache" rule, and generally BSD is fine, but I need to read through it again and make sure.

So, basically it's complicated, and everyone just wants me to slap a license on it, but I'm not comfortable doing that unless I believe it is correct.

**Comment by pombredanne** [comment]
@willnorris re:

> wouldn't this simply inherit the BSD-3-Clause license from Snowball? 

That would be my take and would be the simplest as this is after all derived from the BSD-licensed code Snowball upstream.

@mschoch re:

> So, basically it's complicated, and everyone just wants me to slap a license on it, but I'm not comfortable doing that unless I believe it is correct.

Fair enough but it seems that you are questioning the license of contributions made to Snowball proper and you think these may not have been made under the same license as Snowball proper?
If that's the case, then I suggest that you effectively remove the compiler generated files and leave here the script and instructions to regenerate these, that way you leave your users in command to determine whether they agree with Snowball assertion of its own license?

Also that still leaves open the question of the license to the other code in that repo that is not compiled stemmer code. What would it be? As I guess you could slap a license on that part without any second thoughts, right? 

As far as I am concerned there is a pretty decent trace from the main authors that goes back to summer 2002 which is about as old as it can be since that algorithm and approach to stemming seems to have been invented circa winter 2001. I can seen no indication that there are any other license but BSD-3-Clause that would apply so I would be comfy to treat this as BSD-3-Clause-licensed.

**Comment by mschoch** [comment]
> That would be my take and would be the simplest as this is after all derived from the BSD-licensed code Snowball upstream.

Right, that's where we disagree.  Having written the generator, I don't believe the generated code is any sense "derived" from the generator itself.  That is why we have lawyers.

> If that's the case, then I suggest that you effectively remove the compiler generated files and leave here the script and instructions to regenerate these, that way you leave your users in command to determine whether they agree with Snowball assertion of its own license?

In fact no one needs this repository at all.  I contributed the go generator and runtime upstream here: https://github.com/snowballstem/snowball/tree/master/go  The README documents how to generate them yourself.  I only hosted the generated stemmers (and a copy of the runtime) for convenience.  And because I had a bad experience with the upstream maintainers and wasn't sure the code would actually live there forever.

**Comment by mschoch** [comment]
I will seek additional opinions, I suppose if the snowball project believes all the algorithms are also BSD-3 then I'm fine.  That was not clear to me at the time I originally did this.

**Comment by mschoch** [comment]
OK, I've run this by a couple other people, and we feel comfortable using BSD-3-Clause on this repo, based on the generator and the language files being distributed under the same license.

**Comment by teeler** [comment]
> In fact no one needs this repository at all.

Erm, hold up on that one - this package is included all across the analysis/lang subdirs 
 in http://github.com/blevesearch/bleve, eg: 

https://github.com/blevesearch/bleve/blob/master/analysis/lang/ru/stemmer_ru.go#L21

(https://github.com/blevesearch/bleve/search?p=2&q=snowballstem&unscoped_q=snowballstem)

Am I missing something?

**Comment by mschoch** [comment]
> Erm, hold up on that one - this package is included all across the analysis/lang subdirs

Of course, in reality, having it disappear would could problems for projects already using, including Bleve.  What I meant was, no one is forced to use this repo to use this functionality.  Bleve could have removed support for them directly, and users could have built their own versions and then registered them themselves, and no one would have lost functionality.  It just would have been inconvenient.

Fortunately, no one has to do this now.  I will add the license tonight.

**Comment by teeler** [comment]
Thank you!
